front/src/HOC/withLogger.test.js

```diff
@@ -0,0 +1,211 @@
+      }
+      return true;
+    }
+    componentDidUpdate() {
+      if (this.props.log) {
+        this.props.log('inner component updated');
+      }
+    }
+    componentWillUnmount() {
+      if (this.props.log) {
+        this.props.log('inner component will unmount');
+      }
+    }
+    render() {
+      return <></>;
+    }
+  };
+  ComponentForTest.propsTypes = {
+    log: PropTypes.func,
+  };
+  ComponentForTest.defaultProps = {
+    log: () => {},
+  };
+  return ComponentForTest;
+};
+const fakeLogger = jest.fn(data => data);
+
+describe('Testing withLogger', () => {
+  afterEach(() => {
+    fakeLogger.mockClear();
+  });
+  test('should add log function', () => {
+    const value = withLogger();
+    expect(value).toBeFunction;
+  });
+  test('should add log function with specified params', () => {
+    const WrappedComponent = withLogger({
+      output: fakeLogger,
+      lifecycles: [
+        'didMount',
+        'shouldUpdate',
+        'didUpdate',
+        'willUnmount',
+      ],
+    })(TestComponent);
+    const { rerender, unmount } = render(<WrappedComponent />);
+    expect(fakeLogger).toHaveBeenCalledTimes(1);
+    rerender(<WrappedComponent qw="1" />);
+    expect(fakeLogger).toHaveBeenCalledTimes(3);
+    unmount(<WrappedComponent />);
+    expect(fakeLogger).toHaveBeenCalledTimes(4);
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      1,
+      'component mounted'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      2,
+      'component should update'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      3,
+      'component updated'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      4,
+      'component will unmount'
+    );
+  });
+  test('should add log function with specified HOC name', () => {
+    const WrappedComponent = withLogger()(TestComponent);
+    expect(WrappedComponent.displayName).toBe(
+      'withLogger(TestComponent)'
+    );
+  });
+  test('should log specified lifecycles', () => {
+    const WrappedComponent = withLogger({
+      output: fakeLogger,
+      lifecycles: ['didMount', 'willUnmount'],
+    })(TestComponent);
+    const { unmount } = render(<WrappedComponent />);
+    expect(fakeLogger).toHaveBeenCalledTimes(1);
+    unmount(<WrappedComponent />);
+    expect(fakeLogger).toHaveBeenCalledTimes(2);
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      1,
+      'component mounted'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      2,
+      'component will unmount'
+    );
+  });
+  test('should log props', () => {
+    const WrappedComponent = withLogger({
+      output: fakeLogger,
+      lifecycles: ['props'],
+    })(TestComponent);
+    const {} = render(<WrappedComponent q="1" w="erty" />);
+    expect(fakeLogger).toHaveBeenCalledTimes(1);
+    expect(fakeLogger).toHaveBeenNthCalledWith(1, {
+      q: '1',
+      w: 'erty',
+    });
+  });
+  test('should log props first', () => {
```

Comment:
If I understand the test correctly this code is testing if the logger called twice for each lifecycle event in case there are props passed to the component? And in the test above I see it is called once when there are no props passed. What are you feelings about such an interface? Is it easy and convenient to consume? Is it easy to alter or extend?

Reply:
Not certainly in that way, this test checks the order of logs, which depends on lifecycles param. It will logs twice if we'll push 'props' in array of lifecycles. There is no difference if props are passed to component or not, default value is {} anyway. And log function is called.